geopmdpy/geopmdpy/session.py

```diff
@@ -168,15 +191,27 @@ def run(self, run_time, period, pid, print_header,
                                printed.
 
         """
-        requests = ReadRequestQueue(request_stream)
+        self._requests = ReadRequestQueue(request_stream)
+        do_stats = report_stream is not None
         self.check_read_args(run_time, period)
-        if print_header:
-            header_names = [f'"{name}-{topo.domain_name(domain)}-{domain_idx}"'
-                            if topo.domain_name(domain) != 'board' else f'"{name}"'
-                            for name, domain, domain_idx in requests]
-            print(','.join(header_names), file=out_stream)
-        self.run_read(requests, run_time, period, pid, out_stream)
-
+        if out_stream is not None and print_header:
+            header_names = self.header_names()
+            print(self._delimiter.join(header_names), file=out_stream)
+        with stats.Collector(self._requests) if do_stats else nullcontext() as stats_collector:
+            self.run_read(self._requests, run_time, period, pid, out_stream, stats_collector)
+            if do_stats:
+                report = stats_collector.report_yaml()
+        rank = 0
+        if do_stats and mpi_comm is not None:
+            report_list = mpi_comm.gather(report, root=0)
+            rank = mpi_comm.Get_rank()
+            if rank == 0:
+                tmp_list = [report_list[0]]
+                report_list = [rr.replace('Hosts:\n', '') for rr in report_list[1:]]
```

Comment:
It seems brittle to modify already-formatted report strings in place.

I guess we could avoid the back-and-forth between strings and their structured interpretations by gathering the reports as `dict`s, then build a final report like `dict(host={report_host: report_host_data for report in report_list for report_host, report_host_data in report['host'].items()})`, then have pyyaml dump that dict to `report_stream`. But if we always plan to keep the report strings formatted this way, then that might be unnecessary.

Reply:
resolved with --- separator.